### PR TITLE
feat(manager/dockerfile): add support for registryAliases

### LINKED ISSUE #16855: Allow configuring multiple docker registries using "matchRegistryURL"
### What would you like Renovate to be able to do?

As described in the following discussion: https://github.com/renovatebot/renovate/discussions/16847 I asked whether there is a way to configure multiple docker registries independantly to support image mirroring in kubernetes.

I want this feature to support a kubernetes infrastructure that use mirroring for image registries, and to avoid leaking lookups when renovate goes across private images.

### If you have any ideas on how this should be implemented, please tell us here.

I would like to be able to create a `matchDatasources` based on the registry url to allow mirroring to a private registry. This could be done with a `matchRegistryURL` parameter.

How the implementation could look:

```
{
    "packageRules": [
        {
            "matchDatasources": [
                "docker"
            ],
            "matchRegistryURL": "quay.io"
            "registryUrls": [
                "https://my-quay-mirror.registry.com"
            ]
        },
        {
            "matchDatasources": [
                "docker"
            ],
            "matchRegistryURL": "docker.io"
            "registryUrls": [
                "https://my-docker-mirror.registry.com"
            ]
        }
    ]
}
```

This would allow defining mirrors for independant docker registries by matching the registry url used in the kubernetes manifest and then redirect it to the registry defined in `registryUrls`.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by sockel12**
## Changes

This PR adds support for registryAliases for all docker based managers.

## Context

Hey this is my first PR to this project, could you give me some feedback?
- Closes #16855 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/dockerfile/extract.ts [hunk 1]
```diff
@@ -171,7 +171,9 @@ export function getDep(
     )?.groups;
     if (groups) {
       const dep = {
-        ...getDep(`${value}/${groups.depName}`),
+        ...getDep(
+          `${value}/${groups.depName}` /*, specifyReplaceString, registryAliases (This may be favorable but can lead to loops)*/
```
Comment (viceice) [inline lib/modules/manager/dockerfile/extract.ts:175]: revert this comment, it's unclear what this means

File: lib/modules/manager/dockerfile/extract.ts [hunk 2]
```diff
@@ -322,7 +328,7 @@ export function extractPackageFile(content: string): PackageFile | null {
       } else if (fromImage && stageNames.includes(fromImage)) {
         logger.debug({ image: fromImage }, 'Skipping alias FROM');
       } else {
-        const dep = getDep(fromImage);
+        const dep = getDep(fromImage, undefined, config.registryAliases);
```
Comment (viceice) [inline lib/modules/manager/dockerfile/extract.ts:331]: there are some more `getDep` calls below, which should behave same

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts [hunk 3]
```diff
@@ -822,7 +822,13 @@ const options: RenovateOptions[] = [
       type: 'string',
       format: 'uri',
     },
-    supportedManagers: ['helm-requirements', 'helmv3', 'helmfile', 'gitlabci'],
+    supportedManagers: [
```
Comment (alexander-bruun) [inline lib/config/options/index.ts:825]: Could `docker-compose`, `gitlabci` and `kubernetes` managers also be added here? (if not just all managers under the docker bullet point from the docs)

Comment (viceice) [reply lib/config/options/index.ts:825]: they don't support registry aliases yet, needs seperate PR'S

Comment (alexander-bruun) [reply lib/config/options/index.ts:825]: @viceice Then this pr won't solve my issue, could you mark it as such?

Comment (sockel12) [reply lib/config/options/index.ts:825]: We are currently implementing all the other managers.

File: lib/modules/manager/ansible/__fixtures__/main3.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  docker_container:
+    name: myredis
+    image: quay.io/redis:0.0.1
+    command: redis-server --appendonly yes
+    state: present
+    recreate: yes
+    exposed_ports:
+      - 6379
+    volumes_from:
+      - mydata
```
Comment (viceice) [inline lib/modules/manager/ansible/__fixtures__/main3.yaml:12]:
```suggestion
```

and inline other

File: lib/modules/manager/ansible/extract.ts [hunk 5]
```diff
@@ -2,17 +2,21 @@ import { logger } from '../../../logger';
...
 import * as dockerVersioning from '../../versioning/docker';
 import { getDep } from '../dockerfile/extract';
-import type { PackageDependency, PackageFile } from '../types';
+import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
 
-export function extractPackageFile(content: string): PackageFile | null {
+export function extractPackageFile(
+  content: string,
+  _filename = '',
+  config: ExtractConfig = {}
```
Comment (viceice) [inline lib/modules/manager/ansible/extract.ts:10]:
```suggestion
  _filename,
  config: ExtractConfig 
```

no default values, fix tests

File: lib/modules/manager/ansible/extract.ts [hunk 6]
```diff
@@ -2,17 +2,21 @@ import { logger } from '../../../logger';
...
+): PackageFile | null {
   logger.trace('ansible.extractPackageFile()');
   let deps: PackageDependency[] = [];
   const re = regEx(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
   for (const line of content.split(newlineRegex)) {
     const match = re.exec(line);
     if (match) {
       const currentFrom = match[1];
-      const dep = getDep(currentFrom);
+      const dep = getDep(currentFrom, undefined, config.registryAliases);
```
Comment (viceice) [inline lib/modules/manager/ansible/extract.ts:19]:
```suggestion
      const dep = getDep(currentFrom, undefined, config.registryAliases);
```
```suggestion
      const dep = getDep(currentFrom, true, config.registryAliases);
```

not undefined

File: lib/modules/manager/docker-compose/__fixtures__/docker-compose.3-registry-alias.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+version: "3"
```
Comment (viceice) [inline lib/modules/manager/docker-compose/__fixtures__/docker-compose.3-registry-alias.yml:1]: inline small fixtures

File: lib/modules/manager/docker-compose/extract.ts [hunk 8]
```diff
@@ -30,7 +30,8 @@ class LineMapper {
 
 export function extractPackageFile(
   content: string,
-  fileName?: string
+  fileName?: string,
+  extractConfig: ExtractConfig = {}
```
Comment (viceice) [inline lib/modules/manager/docker-compose/extract.ts:34]:
```suggestion
  fileName: string,
  extractConfig: ExtractConfig
```
see other comment

File: lib/modules/manager/docker-compose/extract.ts [hunk 9]
```diff
@@ -71,7 +72,11 @@ export function extractPackageFile(
     const deps = Object.values(services || {})
       .filter((service) => is.string(service?.image) && !service?.build)
       .map((service) => {
-        const dep = getDep(service.image);
+        const dep = getDep(
+          service.image,
+          undefined,
```
Comment (viceice) [inline lib/modules/manager/docker-compose/extract.ts:77]:
```suggestion
          true,
```

File: lib/modules/manager/dockerfile/extract.ts [hunk 10]
```diff
@@ -226,7 +226,11 @@ export function getDep(
   return dep;
 }
 
-export function extractPackageFile(content: string): PackageFile | null {
+export function extractPackageFile(
+  content: string,
+  _filename = '',
+  config: ExtractConfig = {}
```
Comment (viceice) [inline lib/modules/manager/dockerfile/extract.ts:232]:
```suggestion
  _filename,
  config: ExtractConfig
```

File: lib/modules/manager/dockerfile/extract.ts [hunk 11]
```diff
@@ -322,7 +326,7 @@ export function extractPackageFile(content: string): PackageFile | null {
       } else if (fromImage && stageNames.includes(fromImage)) {
         logger.debug({ image: fromImage }, 'Skipping alias FROM');
       } else {
-        const dep = getDep(fromImage);
+        const dep = getDep(fromImage, undefined, config.registryAliases);
```
Comment (viceice) [inline lib/modules/manager/dockerfile/extract.ts:329]:
```suggestion
        const dep = getDep(fromImage, true, config.registryAliases);
```

File: lib/modules/manager/dockerfile/extract.ts [hunk 12]
```diff
@@ -350,7 +354,11 @@ export function extractPackageFile(content: string): PackageFile | null {
           'Skipping alias COPY --from'
         );
       } else if (Number.isNaN(Number(copyFromMatch.groups.image))) {
-        const dep = getDep(copyFromMatch.groups.image);
+        const dep = getDep(
+          copyFromMatch.groups.image,
+          undefined,
```
Comment (viceice) [inline lib/modules/manager/dockerfile/extract.ts:359]:
```suggestion
          true,
```

File: lib/modules/manager/droneci/extract.ts [hunk 13]
```diff
@@ -1,9 +1,13 @@
...
 import { newlineRegex, regEx } from '../../../util/regex';
 import { getDep } from '../dockerfile/extract';
-import type { PackageDependency, PackageFile } from '../types';
+import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
 
-export function extractPackageFile(content: string): PackageFile | null {
+export function extractPackageFile(
+  content: string,
+  _filename = '',
+  config: ExtractConfig = {}
```
Comment (viceice) [inline lib/modules/manager/droneci/extract.ts:9]:
```suggestion
  _filename,
  config: ExtractConfig
```

File: lib/modules/manager/droneci/extract.ts [hunk 14]
```diff
@@ -34,7 +38,11 @@ export function extractPackageFile(content: string): PackageFile | null {
               currentFrom += finalLineMatch.groups.currentFrom;
               replaceString += '\n' + finalLineMatch.groups.replaceString;
 
-              const dep = getDep(currentFrom);
+              const dep = getDep(
+                currentFrom,
+                undefined,
```
Comment (viceice) [inline lib/modules/manager/droneci/extract.ts:43]:
```suggestion
                true,
```

File: lib/modules/manager/droneci/extract.ts [hunk 15]
```diff
@@ -54,7 +62,11 @@ export function extractPackageFile(content: string): PackageFile | null {
           /^\s* image:\s*'?"?(?<currentFrom>[^\s'"]+)'?"?\s*$/
         ).exec(line);
         if (match?.groups) {
-          const dep = getDep(match.groups.currentFrom);
+          const dep = getDep(
+            match.groups.currentFrom,
+            undefined,
```
Comment (viceice) [inline lib/modules/manager/droneci/extract.ts:67]:
```suggestion
            true,
```

File: lib/modules/manager/kubernetes/extract.ts [hunk 16]
```diff
@@ -3,12 +3,13 @@ import { loadAll } from 'js-yaml';
...
 import { getDep } from '../dockerfile/extract';
-import type { PackageDependency, PackageFile } from '../types';
+import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
 import type { KubernetesConfiguration } from './types';
 
 export function extractPackageFile(
   content: string,
-  fileName: string
+  fileName: string,
+  config: ExtractConfig = {}
```
Comment (viceice) [inline lib/modules/manager/kubernetes/extract.ts:12]:
```suggestion
  config: ExtractConfig
```

File: lib/modules/manager/kubernetes/extract.ts [hunk 17]
```diff
@@ -20,21 +21,24 @@ export function extractPackageFile(
...
+  config: ExtractConfig
+): PackageDependency[] {
   const deps: PackageDependency[] = [];
 
   for (const line of content.split(newlineRegex)) {
     const match = regEx(/^\s*-?\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/).exec(line);
     if (match) {
       const currentFrom = match[1];
-      const dep = getDep(currentFrom);
+      const dep = getDep(currentFrom, undefined, config.registryAliases);
```
Comment (viceice) [inline lib/modules/manager/kubernetes/extract.ts:41]:
```suggestion
      const dep = getDep(currentFrom, true, config.registryAliases);
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts [hunk 18]
```diff
@@ -822,7 +822,17 @@ const options: RenovateOptions[] = [
...
+      'helm-requirements',
+      'helmv3',
+      'helmfile',
+      'gitlabci',
```
Comment (alexander-bruun) [inline lib/config/options/index.ts:829]: Will `gitlabci` manger work if it isn't added in the modules/manager folder?

Comment (sockel12) [reply lib/config/options/index.ts:829]: `gitlabci` was already implemented, so it should work.

**Review by viceice — APPROVED** [review]

**Pull request merged**